### ISSUE TITLE
Slices are overly complicated

### ISSUE THREAD

**Issue by inumanag** [seq-lang/seq#120]
Currently, we have 7 slice classes (`rslice`, `esslice` etc.) for all possible one-dimensional slice combinations. An user has to add a method for each slicing combination. Currently, many internal classes lack such methods.

However, this becomes impractical with multi-dimensional matrices. For example, if you want to implement a matrix that supports `a[1, :2]`, `a[:3, 2:]` and so on, you end with 60+ methods. And good luck with three dimensional matrices.

Can this be resolved with optional parameters? Or it needs more robust type checker?

**Comment by arshajii** [comment]
Maybe we can consolidate to a single slice type?

```python
type slice(start: optional[int], stop: optional[int], step: optional[int])
```

Something like this?

**Comment by inumanag** [comment]
Resolved in #132 . There is new `Slice[start: Optional[int], stop: Optional[int], step: Optional[int])` for this purpose.
